File: java/BaseActivity.java
import android.content.Intent;
import android.support.design.widget.Snackbar;
import android.view.View;

import dagger.android.support.DaggerAppCompatActivity;

public abstract class BaseActivity<V extends BaseAndroidViewModel> extends DaggerAppCompatActivity {

    protected V viewModel;


    @Override
    protected void onStart() {
        super.onStart();
        viewModel.getSnackbarMessage().observe(this,
                (SnackbarMessage.SnackbarObserver) message -> Snackbar.make(getRoot(), message, Snackbar.LENGTH_SHORT).show());
        viewModel.getFinishEvent().observe(this, (Finish.FinishObserver) clazz -> {
            startActivity(new Intent(BaseActivity.this, clazz));
            finish();
        });
    }


    protected abstract View getRoot();
}

File: kotlin/BaseActivity.kt
import android.content.Intent
import android.support.design.widget.Snackbar
import android.view.View

import dagger.android.support.DaggerAppCompatActivity

abstract class BaseActivity<V : BaseAndroidViewModel> : DaggerAppCompatActivity() {
    protected lateinit var viewModel: V


    protected abstract val root: View


    override fun onStart() {
        super.onStart()
        viewModel.snackbarMessage.observe(this,
                { message: String -> Snackbar.make(root, message, Snackbar.LENGTH_SHORT).show() } as SnackbarMessage.SnackbarObserver)
        viewModel.finishEvent.observe(this, { clazz: Class<Any> ->
            startActivity(Intent(this@BaseActivity, clazz))
            finish()
        } as Finish.FinishObserver)
    }
}

